File: python_base_04/core/recall_game/models/player.py
"""
Player Models for Recall Game

This module defines the player system for the Recall card game,
including human players and computer players with AI logic.
"""

from typing import List, Dict, Any, Optional
from enum import Enum
from .card import Card


class PlayerType(Enum):
    """Player types"""
    HUMAN = "human"
    COMPUTER = "computer"


class Player:
    """Base player class for the Recall game"""
    
    def __init__(self, player_id: str, player_type: PlayerType, name: str):
        self.player_id = player_id
        self.player_type = player_type
        self.name = name
        self.hand = []  # 4 cards face down
        self.visible_cards = []  # Cards player has looked at
        self.points = 0
        self.cards_remaining = 4
        self.is_active = True
        self.has_called_recall = False
        self.last_action_time = None
    
    def add_card_to_hand(self, card: Card):
        """Add a card to the player's hand"""
        card.owner_id = self.player_id
        self.hand.append(card)
        self.cards_remaining = len(self.hand)
    
    def remove_card_from_hand(self, card_id: str) -> Optional[Card]:
        """Remove a card from the player's hand"""
        for i, card in enumerate(self.hand):
            if card.card_id == card_id:
                removed_card = self.hand.pop(i)
                self.cards_remaining = len(self.hand)
                return removed_card
        return None
    
    def look_at_card(self, card_id: str) -> Optional[Card]:
        """Look at a specific card in hand"""
        for card in self.hand:
            if card.card_id == card_id:
                card.is_visible = True
                if card not in self.visible_cards:
                    self.visible_cards.append(card)
                return card
        return None
    
    def get_visible_cards(self) -> List[Card]:
        """Get cards that the player has looked at"""
        return [card for card in self.hand if card.is_visible]
    
    def get_hidden_cards(self) -> List[Card]:
        """Get cards that the player hasn't looked at"""
        return [card for card in self.hand if not card.is_visible]
    
    def calculate_points(self) -> int:
        """Calculate total points from cards in hand"""
        return sum(card.points for card in self.hand)
    
    def call_recall(self):
        """Player calls Recall to end the game"""
        self.has_called_recall = True
    
    def can_play_out_of_turn(self, played_card: Card) -> List[Card]:
        """Get cards that can be played out of turn"""
        return [card for card in self.hand if card.can_play_out_of_turn(played_card)]
    
    def to_dict(self) -> Dict[str, Any]:
        """Convert player to dictionary representation"""
        return {
            "player_id": self.player_id,
            "player_type": self.player_type.value,
            "name": self.name,
            "hand": [card.to_dict() for card in self.hand],
            "visible_cards": [card.to_dict() for card in self.visible_cards],
            "points": self.points,
            "cards_remaining": self.cards_remaining,
            "is_active": self.is_active,
            "has_called_recall": self.has_called_recall
        }
    
    @classmethod
    def from_dict(cls, data: Dict[str, Any]) -> 'Player':
        """Create player from dictionary"""
        player_type = PlayerType(data["player_type"])
        player = cls(data["player_id"], player_type, data["name"])
        
        # Restore hand
        for card_data in data.get("hand", []):
            card = Card.from_dict(card_data)
            player.add_card_to_hand(card)
        
        # Restore visible cards
        for card_data in data.get("visible_cards", []):
            card = Card.from_dict(card_data)
            player.visible_cards.append(card)
        
        player.points = data.get("points", 0)
        player.cards_remaining = data.get("cards_remaining", 4)
        player.is_active = data.get("is_active", True)
        player.has_called_recall = data.get("has_called_recall", False)
        
        return player


class HumanPlayer(Player):
    """Human player class"""
    
    def __init__(self, player_id: str, name: str):
        super().__init__(player_id, PlayerType.HUMAN, name)
    
    def make_decision(self, game_state: Dict[str, Any]) -> Dict[str, Any]:
        """Human players make decisions through WebSocket events"""
        # This will be handled by WebSocket events from the frontend
        return {
            "player_id": self.player_id,
            "decision_type": "waiting_for_human_input",
            "available_actions": self._get_available_actions(game_state)
        }
    
    def _get_available_actions(self, game_state: Dict[str, Any]) -> List[str]:
        """Get available actions for the human player"""
        actions = []
        
        if game_state.get("current_player_id") == self.player_id:
            actions.append("play_card")
            actions.append("draw_from_discard")
            actions.append("call_recall")
        
        # Check for out-of-turn plays
        if game_state.get("last_played_card"):
            out_of_turn_cards = self.can_play_out_of_turn(
                game_state["last_played_card"]
            )
            if out_of_turn_cards:
                actions.append("play_out_of_turn")
        
        return actions


class ComputerPlayer(Player):
    """Computer player class with AI decision making"""
    
    def __init__(self, player_id: str, name: str, difficulty: str = "medium"):
        super().__init__(player_id, PlayerType.COMPUTER, name)
        self.difficulty = difficulty
        self.game_logic = self._load_game_logic()
    
    def _load_game_logic(self):
        """Load AI game logic based on difficulty"""
        # This will be implemented to load from YAML files
        from ..game_logic.computer_player_logic import ComputerPlayerLogic
        return ComputerPlayerLogic(self.difficulty)
    
    def make_decision(self, game_state: Dict[str, Any]) -> Dict[str, Any]:
        """Make AI decision based on game state"""
        return self.game_logic.make_decision(game_state, self.to_dict())
    
    def _evaluate_card_value(self, card: Card, game_state: Dict[str, Any]) -> float:
        """Evaluate the value of a card in the current game state"""
        base_value = card.points
        
        # Factor in special powers
        if card.has_special_power():
            base_value -= 2  # Prefer special power cards
        
        # Factor in game progression
        if game_state.get("recall_called"):
            # In final round, minimize points
            return -base_value
        else:
            # During normal play, balance points and utility
            return -base_value * 0.7 + (10 if card.has_special_power() else 0)
    
    def _select_best_card(self, game_state: Dict[str, Any]) -> Optional[Card]:
        """Select the best card to play"""
        if not self.hand:
            return None
        
        # Evaluate all cards
        card_values = []
        for card in self.hand:
            value = self._evaluate_card_value(card, game_state)
            card_values.append((card, value))
        
        # Sort by value (best first)
        card_values.sort(key=lambda x: x[1], reverse=True)
        
        return card_values[0][0] if card_values else None
    
    def _should_call_recall(self, game_state: Dict[str, Any]) -> bool:
        """Determine if the computer should call Recall"""
        if self.has_called_recall:
            return False
        
        # Calculate current position
        total_points = self.calculate_points()
        cards_remaining = len(self.hand)
        
        # Simple AI logic - call Recall if in good position
        if cards_remaining <= 1 and total_points <= 5:
            return True
        
        if cards_remaining <= 2 and total_points <= 3:
            return True
        
        return False 

File: flutter_base_05/lib/modules/recall_game/game_logic/models/player.dart
/// Player Models for Recall Game
///
/// This module defines the player system for the Recall card game,
/// including human players, computer players, and AI decision making.

import 'card.dart';
import '../../../../tools/logging/logger.dart';

// Logging switch for this module
const bool LOGGING_SWITCH = true;

enum PlayerType {
  human,
  computer,
}

enum PlayerStatus {
  waiting,           // Waiting for game to start
  ready,             // Ready to play (waiting for turn)
  playing,           // Currently playing (active turn)
  sameRankWindow,    // Window for out-of-turn same rank plays
  playingCard,       // Player is in the process of playing a card
  drawingCard,       // Player is in the process of drawing a card
  queenPeek,         // Player used queen power to peek at a card
  jackSwap,          // Player used jack power to swap cards
  peeking,           // Player is in peek phase (initial peek, queen peek, etc.)
  initialPeek,       // Player is in initial peek phase (selecting 2 of 4 cards)
  finished,          // Game finished
  disconnected,      // Disconnected from game
  winner,            // Player is the winner of the game
}

class Player {
  /// Base player class for the Recall game
  
  final String playerId;
  final PlayerType playerType;
  final String name;
  
  List<Card?> hand = [];  // 4 cards face down
  List<Card> visibleCards = [];  // Cards player has looked at
  List<Card> knownFromOtherPlayers = [];  // Cards player knows from other players
  int points = 0;
  int cardsRemaining = 4;
  bool isActive = true;
  bool hasCalledRecall = false;
  DateTime? lastActionTime;
  int initialPeeksRemaining = 2;
  List<Card> cardsToPeek = []; // Cards player has peeked at
  PlayerStatus status = PlayerStatus.waiting;  // Player status
  Card? drawnCard;  // Most recently drawn card (Card object)
  
  // Game references for change tracking
  dynamic gameStateManager;
  String? gameId;
  bool _changeTrackingEnabled = true;
  Set<String> _pendingChanges = {};

  Player({
    required this.playerId,
    required this.playerType,
    required this.name,
  });

  void addCardToHand(Card card, {bool isDrawnCard = false, bool isPenaltyCard = false}) {
    card.ownerId = playerId;
    
    // Special handling for drawn cards - always go to the end
    if (isDrawnCard) {
      hand.add(card);
      cardsRemaining = hand.length;
      
      // Manually trigger change detection for hand modification
      _trackChange('hand');
      _sendChangesIfNeeded();
      return;
    }
    
    // For penalty cards and regular cards: look for a blank slot (null) to fill first
    if (isPenaltyCard || !isDrawnCard) {
      for (int i = 0; i < hand.length; i++) {
        if (hand[i] == null) {
          hand[i] = card;
          // Don't update cardsRemaining - we're filling an existing slot
          
          // Manually trigger change detection for hand modification
          _trackChange('hand');
          _sendChangesIfNeeded();
          return;
        }
      }
    }
    
    // If no blank slot found, append to the end
    hand.add(card);
    cardsRemaining = hand.length;
    
    // Manually trigger change detection for hand modification
    _trackChange('hand');
    _sendChangesIfNeeded();
  }

  void setDrawnCard(Card card) {
    drawnCard = card;
    _trackChange('drawnCard');
    _sendChangesIfNeeded();
  }

  Card? getDrawnCard() {
    return drawnCard;
  }

  void clearDrawnCard() {
    drawnCard = null;
    _trackChange('drawnCard');
    _sendChangesIfNeeded();
  }

  Card? removeCardFromHand(String cardId) {
    for (int i = 0; i < hand.length; i++) {
      if (hand[i] != null && hand[i]!.cardId == cardId) {
        final removedCard = hand[i];
        
        // Check if we should create a blank slot or remove the card entirely
        bool shouldCreateBlankSlot = shouldCreateBlankSlotAtIndex(i);
        
        if (shouldCreateBlankSlot) {
          // Replace the card with null (blank slot) to maintain index positions
          hand[i] = null;
        } else {
          // Remove the card entirely and shift remaining cards
          hand.removeAt(i);
        }
        
        // Clear drawn card if the removed card was the drawn card
        if (drawnCard != null && drawnCard!.cardId == cardId) {
          clearDrawnCard();
        }
        
        // Manually trigger change detection for hand modification
        _trackChange('hand');
        _sendChangesIfNeeded();
        
        return removedCard;
      }
    }
    return null;
  }

  void addCardToPeek(Card card) {
    if (!cardsToPeek.contains(card)) {
      cardsToPeek.add(card);
      _trackChange('cardsToPeek');
      _sendChangesIfNeeded();
    }
  }

  void clearCardsToPeek() {
    if (cardsToPeek.isNotEmpty) {
      cardsToPeek.clear();
      _trackChange('cardsToPeek');
      _sendChangesIfNeeded();
    }
  }

  bool shouldCreateBlankSlotAtIndex(int index) {
    // If index is 3 or less, always create a blank slot (maintain initial 4-card structure)
    if (index <= 3) {
      return true;
    }
    
    // For index 4 and beyond, only create blank slot if there are actual cards further up
    // Check if there are any non-null cards at higher indices
    for (int i = index + 1; i < hand.length; i++) {
      if (hand[i] != null) {
        return true;
      }
    }
    
    // No actual cards beyond this index, so remove the card entirely
    return false;
  }

  Card? lookAtCard(String cardId) {
    for (Card? card in hand) {
      if (card != null && card.cardId == cardId) {
        if (!visibleCards.contains(card)) {
          visibleCards.add(card);
        }
        _trackChange('visibleCards');
        _sendChangesIfNeeded();
        return card;
      }
    }
    return null;
  }

  Card? lookAtCardByIndex(int index) {
    if (index >= 0 && index < hand.length && hand[index] != null) {
      final card = hand[index]!;
      if (!visibleCards.contains(card)) {
        visibleCards.add(card);
      }
      _trackChange('visibleCards');
      _sendChangesIfNeeded();
      return card;
    }
    return null;
  }

  List<Card> getVisibleCards() {
    return List.from(visibleCards);
  }

  List<Card> getHiddenCards() {
    return hand.where((card) => card != null && !visibleCards.contains(card)).cast<Card>().toList();
  }

  int calculatePoints() {
    return hand.where((card) => card != null).fold(0, (sum, card) => sum + (card?.points ?? 0));
  }

  void callRecall() {
    hasCalledRecall = true;
    _trackChange('hasCalledRecall');
    _sendChangesIfNeeded();
  }

  void updateStatus(PlayerStatus newStatus) {
    status = newStatus;
    _trackChange('status');
    _sendChangesIfNeeded();
  }

  void setStatus(PlayerStatus newStatus) {
    /// Set player status - alias for updateStatus to match Python version
    updateStatus(newStatus);
  }

  // ========= STATUS CHECKING METHODS =========
  
  bool isPlaying() {
    /// Check if player is currently playing (active turn)
    return status == PlayerStatus.playing;
  }

  bool isReady() {
    /// Check if player is ready (waiting for turn)
    return status == PlayerStatus.ready;
  }

  bool isWaiting() {
    /// Check if player is waiting (game not started)
    return status == PlayerStatus.waiting;
  }

  bool isSameRankWindow() {
    /// Check if player is in same rank window (can play out-of-turn)
    return status == PlayerStatus.sameRankWindow;
  }

  bool isPlayingCard() {
    /// Check if player is in process of playing a card
    return status == PlayerStatus.playingCard;
  }

  bool isDrawingCard() {
    /// Check if player is in process of drawing a card
    return status == PlayerStatus.drawingCard;
  }

  bool isQueenPeek() {
    /// Check if player is in queen peek status (used queen power)
    return status == PlayerStatus.queenPeek;
  }

  bool isJackSwap() {
    /// Check if player is in jack swap status (used jack power)
    return status == PlayerStatus.jackSwap;
  }

  bool isPeeking() {
    /// Check if player is in peeking phase
    return status == PlayerStatus.peeking;
  }

  bool isInitialPeek() {
    /// Check if player is in initial peek phase
    return status == PlayerStatus.initialPeek;
  }

  bool isFinished() {
    /// Check if player has finished the game
    return status == PlayerStatus.finished;
  }

  bool isDisconnected() {
    /// Check if player is disconnected
    return status == PlayerStatus.disconnected;
  }

  bool isWinner() {
    /// Check if player is the winner
    return status == PlayerStatus.winner;
  }

  void updatePoints() {
    points = calculatePoints();
    _trackChange('points');
    _sendChangesIfNeeded();
  }

  void setGameReferences(dynamic gameStateManager, String gameId) {
    this.gameStateManager = gameStateManager;
    this.gameId = gameId;
  }

  void enableChangeTracking() {
    /// Enable automatic change tracking
    _changeTrackingEnabled = true;
  }

  void disableChangeTracking() {
    /// Disable automatic change tracking
    _changeTrackingEnabled = false;
  }

  void _trackChange(String propertyName) {
    if (_changeTrackingEnabled) {
      _pendingChanges.add(propertyName);
    }
  }

  void _sendChangesIfNeeded() {
    if (!_changeTrackingEnabled || _pendingChanges.isEmpty) {
      return;
    }

    if (gameStateManager == null || gameId == null) {
      return;
    }

    try {
      Logger().info('🔄 Player _sendChangesIfNeeded called with ${_pendingChanges.length} pending changes', isOn: LOGGING_SWITCH);
      Logger().info('=== SENDING PLAYER UPDATE ===', isOn: LOGGING_SWITCH);
      Logger().info('Player ID: $playerId', isOn: LOGGING_SWITCH);
      Logger().info('Changed properties: ${_pendingChanges.toList()}', isOn: LOGGING_SWITCH);
      Logger().info('=============================', isOn: LOGGING_SWITCH);
      
      // Get the coordinator from the game state manager
      if (gameStateManager.appManager != null) {
        final coordinator = gameStateManager.appManager.gameEventCoordinator;
        if (coordinator != null) {
          // Send player state update using existing coordinator method
          coordinator.sendPlayerStateUpdate(gameId!, playerId);
          Logger().info('Player update sent successfully for properties: ${_pendingChanges.toList()}', isOn: LOGGING_SWITCH);
          
          // Also trigger GameState players property change detection
          _triggerGamestatePlayersUpdate();
        } else {
          Logger().info('No coordinator found for player update', isOn: LOGGING_SWITCH);
        }
      } else {
        Logger().info('No app_manager found for player update', isOn: LOGGING_SWITCH);
      }
      
      // Clear pending changes
      _pendingChanges.clear();
      
    } catch (e) {
      Logger().error('Error in player _sendChangesIfNeeded: $e', isOn: LOGGING_SWITCH);
    }
  }

  void _triggerGamestatePlayersUpdate() {
    /// Trigger GameState players property change detection to send room-wide update
    try {
      Logger().info('🔄 Triggering GameState players property update for player: $playerId', isOn: LOGGING_SWITCH);
      
      // Get the game state from the game state manager
      try {
        // Use reflection or try-catch to call methods dynamically
        final gameState = gameStateManager.getGame?.call(gameId);
        if (gameState != null) {
          // Try to call the change tracking methods
          gameState._trackChange?.call('players');
          gameState._sendChangesIfNeeded?.call();
          Logger().info('✅ GameState players property update triggered successfully', isOn: LOGGING_SWITCH);
        } else {
          Logger().info('❌ GameState not found', isOn: LOGGING_SWITCH);
        }
      } catch (e) {
        Logger().info('❌ GameStateManager method call failed: $e', isOn: LOGGING_SWITCH);
      }
      
    } catch (e) {
      Logger().error('❌ Error triggering GameState players update: $e', isOn: LOGGING_SWITCH);
    }
  }

  Map<String, dynamic> toDict() {
    return {
      'player_id': playerId,
      'player_type': playerType.name,
      'name': name,
      'hand': hand.map((card) => card?.toDict()).toList(),
      'visible_cards': visibleCards.map((card) => card.toDict()).toList(),
      'known_from_other_players': knownFromOtherPlayers.map((card) => card.toDict()).toList(),
      'points': points,
      'cards_remaining': cardsRemaining,
      'is_active': isActive,
      'has_called_recall': hasCalledRecall,
      'last_action_time': lastActionTime?.toIso8601String(),
      'initial_peeks_remaining': initialPeeksRemaining,
      'cards_to_peek': cardsToPeek.map((card) => card.toDict()).toList(),
      'status': status.name,
      'drawn_card': drawnCard?.toDict(),
    };
  }

  factory Player.fromDict(Map<String, dynamic> data) {
    final player = Player(
      playerId: data['player_id'],
      playerType: PlayerType.values.firstWhere((e) => e.name == data['player_type']),
      name: data['name'],
    );
    
    player.points = data['points'] ?? 0;
    player.cardsRemaining = data['cards_remaining'] ?? 4;
    player.isActive = data['is_active'] ?? true;
    player.hasCalledRecall = data['has_called_recall'] ?? false;
    player.initialPeeksRemaining = data['initial_peeks_remaining'] ?? 2;
    player.status = PlayerStatus.values.firstWhere((e) => e.name == data['status'], orElse: () => PlayerStatus.waiting);
    
    if (data['last_action_time'] != null) {
      player.lastActionTime = DateTime.parse(data['last_action_time']);
    }
    
    if (data['drawn_card'] != null) {
      player.drawnCard = Card.fromDict(data['drawn_card']);
    }
    
    return player;
  }
}

class HumanPlayer extends Player {
  HumanPlayer({
    required String playerId,
    required String name,
  }) : super(
    playerId: playerId,
    playerType: PlayerType.human,
    name: name,
  );

  Map<String, dynamic> makeDecision(Map<String, dynamic> gameState) {
    /// Human players make decisions through WebSocket events
    // This will be handled by WebSocket events from the frontend
    return {
      'player_id': playerId,
      'decision_type': 'waiting_for_human_input',
      'available_actions': _getAvailableActions(gameState),
    };
  }

  List<String> _getAvailableActions(Map<String, dynamic> gameState) {
    /// Get available actions for the human player
    final actions = <String>[];
    
    if (gameState['current_player_id'] == playerId) {
      actions.add('play_card');
      actions.add('draw_from_discard');
      actions.add('call_recall');
    }
    
    // Check for out-of-turn plays
    if (gameState['last_played_card'] != null) {
      // This would need to be implemented based on game rules
      // For now, just add the action if there's a last played card
      actions.add('play_out_of_turn');
    }
    
    return actions;
  }
}

class ComputerPlayer extends Player {
  final String difficulty;
  
  ComputerPlayer({
    required String playerId,
    required String name,
    this.difficulty = "medium",
  }) : super(
    playerId: playerId,
    playerType: PlayerType.computer,
    name: name,
  );

  Map<String, dynamic> makeDecision(Map<String, dynamic> gameState) {
    /// Make AI decision based on game state using built-in logic
    
    // Use built-in AI logic methods
    final bestCard = _selectBestCard(gameState);
    final shouldCallRecall = _shouldCallRecall(gameState);
    
    if (shouldCallRecall) {
      return {
        'action': 'call_recall',
        'reason': 'AI decided to call recall (difficulty: $difficulty)',
        'player_id': playerId,
      };
    }
    
    if (bestCard != null) {
      // Find the card index in hand
      final cardIndex = hand.indexWhere((card) => card != null && card.cardId == bestCard.cardId);
      return {
        'action': 'play_card',
        'card_index': cardIndex != -1 ? cardIndex : 0,
        'reason': 'AI selected best card (difficulty: $difficulty)',
        'player_id': playerId,
      };
    }
    
    // Fallback: play first card
    return {
      'action': 'play_card',
      'card_index': 0,
      'reason': 'AI fallback decision (difficulty: $difficulty)',
      'player_id': playerId,
    };
  }

  double _evaluateCardValue(Card card, Map<String, dynamic> gameState) {
    /// Evaluate the value of a card in the current game state
    double baseValue = card.points.toDouble();
    
    // Factor in special powers
    if (card.hasSpecialPower()) {
      baseValue -= 2; // Prefer special power cards
    }
    
    // Factor in game progression
    if (gameState['recall_called'] == true) {
      // In final round, minimize points
      return -baseValue;
    } else {
      // During normal play, balance points and utility
      return -baseValue * 0.7 + (card.hasSpecialPower() ? 10 : 0);
    }
  }

  Card? _selectBestCard(Map<String, dynamic> gameState) {
    /// Select the best card to play
    final validCards = hand.where((card) => card != null).cast<Card>().toList();
    if (validCards.isEmpty) {
      return null;
    }
    
    // Evaluate all cards
    final cardValues = <MapEntry<Card, double>>[];
    for (final card in validCards) {
      final value = _evaluateCardValue(card, gameState);
      cardValues.add(MapEntry(card, value));
    }
    
    // Sort by value (best first)
    cardValues.sort((a, b) => b.value.compareTo(a.value));
    
    return cardValues.isNotEmpty ? cardValues.first.key : null;
  }

  bool _shouldCallRecall(Map<String, dynamic> gameState) {
    /// Determine if the computer should call Recall
    if (hasCalledRecall) {
      return false;
    }
    
    // Calculate current position
    final totalPoints = calculatePoints();
    final cardsRemaining = hand.where((card) => card != null).length;
    
    // Simple AI logic - call Recall if in good position
    if (cardsRemaining <= 1 && totalPoints <= 5) {
      return true;
    }
    
    if (cardsRemaining <= 2 && totalPoints <= 3) {
      return true;
    }
    
    return false;
  }

  void updateKnownFromOtherPlayers(Card card, Map<String, dynamic> gameState) {
    /// Update the player's known cards from other players list
    if (!knownFromOtherPlayers.any((c) => c.cardId == card.cardId)) {
      knownFromOtherPlayers.add(card);
    }
  }
}